In [49]:
import pandas as pd
import numpy as np
%matplotlib inline
import otter
grader = otter.Notebook()

# A1 | Week 3: Practice While Loops

------

### Instructions:

- Complete 50 points worth of questions to pass the assessment
  - You can attempt any number of questions and in any order provided you pass at least 50 points.
- Develop and check your answers in the spaces provided.
  - Do not remove any variables other provided markings already present in the answer spaces.
  - Do not make any changes to this notebook outside of the answer spaces provided.
  
#### Submitting

- Upload your .ipynb file to Gradescope (Make sure to restart the kernel and run all before you submit)
- You can submit as many times as needed
- When reviewing results, ignore any results listed under "Public Tests"

For more information, see the assessment page. We have also provided an example question below for clarity. 

#### Example Question
Write a Python program to display the first 10 integer numbers starting with 1 using while loops. Create the variable `example_output` and use it inside the while loop to construct the string that you want to display at the end of the program.

```
#This solution is for demonstration purposes ONLY. 
#There may be other ways of doing this task using while loops that produce the same output.
#You have the freedom to use a solution that you prefer as long as it uses while loop and 
#produces the expected output using the example_output variable.
example_output = ""
example_count = 1
while example_count <= 10:
    example_output += str(example_count)
    example_output += "\n"
    example_count += 1
example_output = example_output.strip() #this is to remove the extra new line in 10
print(example_output)
```

#### Question 1&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(1 Point)
Write a Python program to display integer numbers starting with 190 to 200 using while loops. Create the variable `output1` and use it inside the while loop to construct the string that you want to display at the end of the program.

Expected Output:   
190  
191  
192  
193  
194  
195  
196  
197  
198  
199  
200

In [145]:
i=190
output1=""
while (i<=200):
    
    output1 += str(i)+"\n"
    i=i+1
# Write your solution here
output1=output1.strip()
output1

'190\n191\n192\n193\n194\n195\n196\n197\n198\n199\n200'

#### Question 2&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(1 Point)
Write a Python program to display the first 10 even numbers starting with 0 using while loops. Create the variable `output2` and use  it inside the while loop to construct the string that you want to display at the end of the program.

*Note: The displayed string should be in the same format as in Q1. For clarity, the expected output is shown below.*

Expected Output:
<p>
0<br> 
2<br> 
4<br> 
6<br> 
8<br> 
10<br> 
12<br>
14<br> 
16<br> 
18
</p>

In [51]:
i=0

output2=""
while(i<20):

    if i%2==0:
        output2+=str(i)+"\n"
    i=i+1
# Write your solution here
output2=output2.strip()
print(output2)

0
2
4
6
8
10
12
14
16
18



#### Question 3&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(1 Point)
Write a Python program to display the following series using while loops: `30, 40, 50, 60, 70, 80, 90, ..., 300`. Create the variable `output3` and use  it inside the while loop to construct the string that you want to display at the end of the program.

*Note: The displayed string should be in the same format as in Q1. For clarity, the expected output is shown below.*

Expected Output:
<p>
30<br> 
40<br> 
50<br> 
60<br> 
70<br> 
80<br> 
90<br> 
...<br> 
300
</p>

In [146]:
i=30

output3=""
while(i<=300):
    output3+=str(i)+"\n"
    i=i+10
    

   

# Write your solution here
output3=output3.strip()
print(output3)

30
40
50
60
70
80
90
100
110
120
130
140
150
160
170
180
190
200
210
220
230
240
250
260
270
280
290
300


#### Question 4&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(1 Point)
Write a Python program to display the squares of integer numbers divisible by 4 until 200 (inclusive) using while loops. Create the variable `output4` and use it inside the while loop to construct the string that you want to display at the end of the program. (Expected output series: 16, ..., 40000)

*Note: The displayed string should be in the same format as in Q1.*

In [53]:
i=4
output4=""
while(i<=200):

    
    output4 +=str(i**2)+"\n"
    i=i+4

# Write your solution here

print(output4)

16
64
144
256
400
576
784
1024
1296
1600
1936
2304
2704
3136
3600
4096
4624
5184
5776
6400
7056
7744
8464
9216
10000
10816
11664
12544
13456
14400
15376
16384
17424
18496
19600
20736
21904
23104
24336
25600
26896
28224
29584
30976
32400
33856
35344
36864
38416
40000



#### Question 5&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(1 Point)
Write a Python program to print the following series using while loops: `120, 116, 112, ..., 36`. Create the variable `output5` and use it inside the while loop to construct the string that you want to display at the end of the program.

*Note: The displayed string should be in the same format as in Q1.*

In [54]:
i=120
output5=""
while(i>=36):

    
    output5 +=str(i)+"\n"
    i-=4



print(output5)

120
116
112
108
104
100
96
92
88
84
80
76
72
68
64
60
56
52
48
44
40
36



#### Question 6&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(3 Points)
Write a Python program that adds all the odd numbers until 300 (inclusive) using while loops. Create the variable `ouput6` and use it inside the while loop for the calculations needed. Print `output6` at the end of the program to check if it works correctly.

In [147]:
output6 = 0
i=1
output6=0
while(i<=300):
    if i%2==1:
        output6+=i
    i+=1
# Write your solution here



#### Question 7&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(3 Points) 
Write a Python program to calculate the total of first 10 integer numbers using while loops. However, if the number is divisible by 5, do not use them to calculate the total value. Create the variable `output7` and use it inside the while loop for the calculations needed. [Expected output: 40]
***
<details>
  <summary><span style="color:blue">Click to see Hints</span></summary>
    
* You will need to use `continue` statement. This statement rejects all the remaining statements in the current iteration of the loop and moves the control back to the top of the loop.
</details>

In [152]:
i=0
sum=0
while(i<10):
    if i%5 !=0:
        sum+=i
    i+=1

print(output7)

0


#### Question 8&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(3 Points) 
Write a Python program to calculate the total of first 20 integer numbers using while loops. However, if the program encounters a number that is divisible by 10 between 1 to 20, stop looping at the very moment and  calculate the total value until that point (i.e., excluding the number that is divisible by 10 and the numbers after that are also excluded). Create the variables `output8` and use it in the program for the calculations needed. [Expected output: 45]
***
<details>
  <summary><span style="color:blue">Click to see Hints</span></summary>
    
* You will need to use `break` statement. This will terminate the current loop.
</details>

In [158]:
i=1
sum=0
output7=0
while(i<20):
    if i%10==0:
       
        break
    output7+=i
    i+=1

print(output7)

45


#### Question 9&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(3 Points)
Write a Python program to display the average (rounded to two decimal places) of all the numbers which are divisible by 13 but not by 3 between 300 and 600.(exclusive both numbers) using while loop. Use the variable `output9` to assign the solution to it.

In [58]:
# Write your solution here



#### Question 10&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(3 Points)
Write a Python program to display the following pattern using while loops. Create the variable `output10` and use it inside the while loop to construct the string that you want to display at the end of the program.
<p>
*<br>
**<br>
***<br>
****<br>
*****<br>
</p>

In [162]:
output10 = ""

# Write your solution here

for i in range(1,6):
    for j in range(i):
        output10+="*"
    output10+="\n"
        
output10=output10.strip()
'*\n**\n***\n****\n*****'

'*\n**\n***\n****\n*****'

#### Question 11&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)
Write a Python program to display the following pattern using while loops. Create the variable `output11` and use it inside the while loop to construct the string that you want to display at the end of the program.
<p>
1<br>
22<br>
333<br>
4444<br>
55555<br>
666666<br>
7777777<br>
88888888<br>
999999999<br>
</p>

In [168]:
output11 = ""
for i in range(1,10):
    for j in range(i):
        output11+=str(i)
    output11+="\n"
        
output11=output11.strip()
# Write your solution here

output11
'1\n22\n333\n4444\n55555\n666666\n7777777\n88888888\n999999999'

'1\n22\n333\n4444\n55555\n666666\n7777777\n88888888\n999999999'

#### Question 12&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)
Write a Python program to display numbers that are divisible by 9 and multiple of 5, between 1000 and 2000 (including 1000 and 2000) using while loops. Create the variable `output12` and use it inside the while loop to construct the string that you want to display at the end of the program.

*Note: The displayed string should be in the same format as in Q1.*

In [61]:
output12 = ""

# Write your solution here



#### Question 13&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)
Write a Python program that iterates from 200 to 250 (including both 200 and 250) using while loops and perform the following. Create the variable `output13` and use it inside the while loop to construct the string that you want to display at the end of the program.

1. If a number is a multiply of 3, display "NUMBER is a multiply of 3!" (*Note: NUMBER should be replaced with the value at that iteration*)
2. If a number is a multiply of 5, display "NUMBER is a multiply of 5!" (*Note: NUMBER should be replaced with the value at that iteration*)
3. If a number is a multiply of 3 as well as multiply of 5, display "NUMBER is a multiply of both 3 and 5!" (*Note: NUMBER should be replaced with the value at that iteration*)

*Note: The displayed string should be in the same format as in Q1.*

In [62]:
output13 = ""

# Write your solution here



#### Question 14&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)
Create the variable `q14_num` and assign 1912 to it. Write a Python program to reverse `q14_num` using while loop and arithmetic operations. For example, if `q14_num` is 1912, the reversed number is 2191. Create the variable `output14` and assign the reversed number to it. However, if a user has mistakenly assign one digit number (i.e., 0-9) to `q14_num`, `output14`should be the string "Please enter at least two digits to perform operations". Print `output14` at the end of the prohram to check if it works correctly.

In [63]:
q14_num = 1912


rev=""
if q14_num>10:
    output14=0
    Reminder=0
    while(q14_num > 0):    
        Reminder = q14_num %10    
        output14 = (output14 *10) + Reminder    
        q14_num = q14_num //10  
elif q14_num<10 :
    output14="Please enter at least two digits to perform operations"


output14

2191

#### Question 15 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points) 
Write a Python program to display the series of `2, 22, 222, 2222, ..., n terms` till n terms using while loops.

For this question, create `q15_n` variable (that represents `n` in the series) and assign 20 to it. Create the variable `output15` and use it inside the while loop to construct the string that you want to display at the end of the program.

*Note: The displayed string should be in the same format as in Q1.*

In [98]:
q15_n = 20
output15=""

for i in range (q15_n):
    for j in range(i+1):
        output15+="2"
    output15+="\n"
    
 
output15=output15[:-1].strip()
output15
# Write your solution here




'2\n22\n222\n2222\n22222\n222222\n2222222\n22222222\n222222222\n2222222222\n22222222222\n222222222222\n2222222222222\n22222222222222\n222222222222222\n2222222222222222\n22222222222222222\n222222222222222222\n2222222222222222222\n22222222222222222222'

#### Question 16 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(4 Points)
Write a Python program to find the sum of following series: 1 + 4 - 9 + 16 - 25 + 36 ... n terms.

For this question, create `q16_n` variable (that represents `n` in the series) and assign 20 to it. Create the variable `output16` and assign the solution to it.

In [65]:
q16_n = 20
output16=0
for i in range(2,q16_n+1):
    
    if i%2==0:
  
        output16+=i**2
    else:
        
        output16-=i**2
        
output16+=1       
output16       
# Write your solution here




212

#### Question 17 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)
 Write a Python program to calculate factorial using while loop. Create a variable `q17_input` and assign the number 191 to it. To calculate the factorial use the following guidelines. Create a variable `output17` and assign the solution to it.

* Factorial of number n is represented by `n!` and it is equal to `1*2*3*....*(n-1)*n`. For example, `4!` = `1*2*3*4` = 24
* Note that 1! = 1 and 0! = 1

In [66]:
q17_input = 191
output17=1

if q17_input==1:
    output17=1
elif q17_input==0:
    output17=1
else :
    i=1
    while(i<=q17_input):
        
        output17=i*output17
       
        i+=1
print(output17)
# Write your solution here




1848941630973752588819559184291528260234868800748408811903414244709323983074218587659329898912306172360114330860464468736986865023838872462695380757512438036901650776605459478797800350707553256022912105863632666857472153971092009065677258893759594079568220432651120989901182064402330506648926574264895332679680000000000000000000000000000000000000000000000


#### Question 18 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)
Write a Python program to display the following pattern using while loops. Create the variable `output18` and use it inside the while loop to construct the string that you want to display at the end of the program.
<p>
    &nbsp;&nbsp;&nbsp;&nbsp;*  <br>
  &nbsp;&nbsp;&nbsp;*** <br>
  &nbsp;&nbsp;***** <br>
&nbsp;******* <br>
********* <br>
&nbsp;******* <br>
&nbsp;&nbsp;***** <br>
   &nbsp;&nbsp;&nbsp;*** <br>
    &nbsp;&nbsp;&nbsp;&nbsp;*  <br>
</p>

***
<details>
  <summary><span style="color:blue">Click to see Hints</span></summary>
    
* You may need to use two while loops to print the first half and second half of the pattern.
* You may need to use `break` statement as practiced in Question 8.
</details>

In [67]:
output18 = ""

# Write your solution here



#### Question 19 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)
Write a Python program to reverse a given number using while loop. Create a variable `q19_input` and assign the number 121019200109 to it. Use the variable `output19` and assign the solution to it.

In [88]:
q19_input = 121019200109
output19=0
Reminder=0

while(q19_input > 0):    
        Reminder = q19_input %10    
        output19 = (output19 *10) + Reminder    
        q19_input = q19_input //10  

output19
# Write your solution here




901002910121

#### Question 20 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)
Write a Python program to find the sum of the following series using a while loop.  
$1 +\frac{y}{1!} + \frac{y^2}{2!} + ... + \frac{y^n}{n!}$

Use the following guidelines when developing the program.

* Create two variables `q20_y` and `q20_n` and assign the values 6 and 26 (Note that `q20_y` and `q20_n` represent `y` and `n` in the above-mentioned series).
* The sum should be ronded to the nearest integer value. Create the variable `output20` and assign the solution to it.

In [69]:
q20_y, q20_n = 6, 26

# Write your solution here



#### Question 21 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)
Write a Python program to check whether a number is palindrome or not using while loop.
Palindrome is a sequence which reads the same backward as forward. Examples of palindromes are mentioned below.
* 12344321
* 2002
* 19191

Create the variable `q21_input` and assign 12123454321 to it. If `q21_input` is a palindrome assign the string "This is a palindrome" to `output21`. Otherwise, `output21` should be "This is not a palindrome"

In [96]:
q21_input = 19191
q21_input_copy=0
i=0
s=""

while q21_input<=0:
    Reminder = q21_input %10    
    q21_input_copy = (q21_input *10) + Reminder    
    q21_input = q21_input //10  
print(q21_input_copy)
if q21_input==q21_input_copy:
    output21="This is a palindrome" 
else:
    output21="This is not a palindrome"     
    
# Write your solution here



0


#### Question 22 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(6 Points)
Write a Python program that displays sum of odd numbers and even numbers separately that fall between two given numbers `q22_num1` and `q22_num2` (including both numbers) using while loop. Use the following guidelines to develop this program.

* For this question, assign the values 21 and 9 to the two variables `q22_num1` and `q22_num2`, respectively.
* Use the two variables `output22_odd` and `output22_even` to assign the sum of odd numbers and the sum of even numbers, respectively.

In [71]:
q22_num1 = 21
q22_num2 = 9

# Write your solution here



#### Question 23 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(10 Points)
Write a Python program to display the following pattern using while loops. Create the variable `output23` and use it inside the while loop to construct the string that you want to display at the end of the program.
<p>
 &nbsp;***<br>                                                                 
 *&nbsp;&nbsp;&nbsp;*<br>                                                                  
 *&nbsp;&nbsp;&nbsp;*<br>                                                                  
 *&nbsp;&nbsp;&nbsp;*<br>                                                                  
 *&nbsp;&nbsp;&nbsp;*<br>                                                                  
 *&nbsp;&nbsp;&nbsp;*<br>                                                                  
&nbsp;*** 
</p>

In [143]:
output23 = "" 
i=0
while (i<7):
   
    if i==0 or i==6:
        output23+="  ***  \n"
    else:
        output23+=" *   * \n"
    i=i+1
    
output23=output23.rstrip()
        
# Write your solution here

output23


'  ***  \n *   * \n *   * \n *   * \n *   * \n *   * \n  ***'

#### Question 24 &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(10 Points)
Write a Python program to display the following pattern using while loops. Create the variable `output24` and use it inside the while loop to construct the string that you want to display at the end of the program.

<p>
    * * * * *<br> 
&nbsp;* * * *<br> 
&nbsp;&nbsp;* * *<br> 
&nbsp;&nbsp;&nbsp;* *<br> 
&nbsp;&nbsp;&nbsp;&nbsp;*<br> 
&nbsp;&nbsp;&nbsp;&nbsp;*<br> 
&nbsp;&nbsp;&nbsp;* *<br> 
&nbsp;&nbsp;* * *<br> 
&nbsp;* * * *<br> 
* * * * *<br>
</p>

In [141]:
output24 = ""

# Write your solution here
out1,out2="",""

f=0
for i in range(5,0,-1):
    x=""
    for k in range(f):
        x+=" "
    f=f+1
    for j in range(i):
        x+="* "
    x+='\n'
    out1+=x
out1

'* * * * * \n * * * * \n  * * * \n   * * \n    * \n'

In [136]:
output=""
for  i in range(1,5):
    for j in range(0,5-i):
        output+=" "
    for k in range(i,0):
        
        output+=k
    output+="\n"
output       

'    \n   \n  \n \n'